File: python/module/ViewControlModule/ViewControlModule.py
# This is a sample Python script.

# Press Shift+F10 to execute it or replace it with your code.
# Press Double Shift to search everywhere for classes, files, tool windows, actions, and settings.

from PyQt5 import QtCore, QtGui, QtQml, QtQuick
from PyQt5.QtCore import QObject, QUrl, pyqtSignal, pyqtSlot, Qt
from PyQt5.QtQml import QQmlApplicationEngine, qmlRegisterType
from PyQt5.Qt import QMetaObject, Q_ARG, Q_RETURN_ARG

from python.res import rc_qml
from python.src.module.BaseModule import QBaseModule
from python.src.utils.logger import getLogger

from python.src.define.viewDefs import *
from .handler.WindowEventHandler import WindowEventHandler
from .handler.ViewEventHandler import ViewEventHandler
from .handler.ListEventHandler import ListEventHandler
from .handler.ButtonEventHandler import ButtonEventHandler
from .handler.ComboBoxEventHandler import ComboBoxEventHandler
from .handler.DialogEventHandler import DialogEventHandler
from .handler.TextEditEventHandler import TextEditEventHandler

class ViewControlModule(QBaseModule):
    _strModultName = 'ViewControlModule'
    __engine = None

    __hWindowEvent = None
    __hViewEvent = None
    __hLstEvent = None
    __hBtnCtlEvent = None
    __hCbxCtlEvent = None
    __hDlgEvent = None
    __hEdtCtlEvent = None

    def __init__(self, parent=None):
        super().__init__(parent)

    def init(self) -> None:
        if self.__initQmlEngine() == False:
            getLogger().error("Failed to init QML Engine")
            return

        if not self.__hWindowEvent:
            self.__hWindowEvent = WindowEventHandler()
        if not self.__hViewEvent:
            self.__hViewEvent = ViewEventHandler()
        if not self.__hLstEvent:
            self.__hLstEvent = ListEventHandler()
        if not self.__hBtnCtlEvent:
            self.__hBtnCtlEvent = ButtonEventHandler()
        if not self.__hCbxCtlEvent:
            self.__hCbxCtlEvent = ComboBoxEventHandler()
        if not self.__hDlgEvent:
            self.__hDlgEvent = DialogEventHandler()
        if not self.__hEdtCtlEvent:
            self.__hEdtCtlEvent = TextEditEventHandler()

        self.SetControlEventConnections()

        super().init()

    @QtCore.pyqtSlot()
    def doRun(self) -> None:
        pass

    @QtCore.pyqtSlot()
    def stopModule(self) -> None:
        self.SetControlEventDisconnections()

        if self.__engine:
            self.__engine.exit(0)
            self.__engine.deleteLater()
            self.__engine = None

        super().stopModule()
        self.__clear()

    def __initQmlEngine(self) -> bool:
        if not self.__engine:
            self.__engine = QQmlApplicationEngine()

            self.__engine.addImportPath('qrc:/')
            self.__engine.addImportPath('qrc:///BaseComponents')
            self.__engine.addImportPath('qrc:///components')

            qmlRegisterType(ViewEnum, 'ViewEnum', 1, 0, 'ViewEnum')
            self.__engine.load('qrc:/main.qml')
            if len(self.__engine.rootObjects()) <= 0:
                getLogger().error('Failed to init engine')
                return False

        return True

    def __clear(self) -> None:
        pass

    def SetControlEventConnections(self):
        objRoot = self.GetRootControlObject()
        if not objRoot:
            getLogger().error('Root object is None')
            return

        nEventConnected = objRoot.property(PROP_IS_EVENT_CONNECTED)
        if not nEventConnected:
            objRoot.sigNewControlLoaded.connect(self.HandleNewControlLoadedEvent, Qt.DirectConnection)
            objRoot.sigDestroyControlLoaded.connect(self.HandleDestroyControlLoadedEvent, Qt.DirectConnection)
            objRoot.sigAppModeChanged.connect(self.HandleAppModeChangedEvent, Qt.DirectConnection)

            appMode = objRoot.property(PROP_N_APP_MODE)
            if appMode != None:
                self.HandleAppModeChangedEvent(appMode)

            objRoot.setProperty(PROP_IS_EVENT_CONNECTED, True)

        self.SetControlEventConnection(objRoot, True)

    def SetControlEventConnection(self, obj, bRecursive):
        if not obj:
            getLogger().error('Object is null')
            return

        if bRecursive:
            self.SetControlEventConnectionRecursive(obj)
        else:
            self.connectObjectEvent(obj)

    def connectObjectEvent(self, obj):
        if not obj:
            getLogger().error('Object is null')
            return

        bConnectEvent = obj.property(PROP_B_CONNECT_EVENT)
        if not bConnectEvent:
            return

        nObjectType = obj.property('objectType')
        if not nObjectType:
            getLogger().error('Object type is null')
            return

        if nObjectType == ViewEnum.eObjectType.OBJECT_WINDOW:
            self.__hWindowEvent.connectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_VIEW:
            self.__hViewEvent.connectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_DIALOG:
            self.__hDlgEvent.connectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_BUTTON_CONTROL:
            self.__hBtnCtlEvent.connectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_COMBOBOX_CONTROL:
            self.__hCbxCtlEvent.connectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_TEXTEDIT_CONTROL:
            self.__hEdtCtlEvent.connectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_LIST_VIEW:
            self.__hLstEvent.connectObjectEvent(obj)
        else: # nObjectType == ViewEnum::OBJECT_NONE or unknown
            getLogger().error('Unknown objectType: %d', nObjectType)

    def SetControlEventDisconnections(self):
        objRoot = self.GetRootControlObject()
        if not objRoot:
            getLogger().error('Root object is None')
            return

        nEventConnected = objRoot.property(PROP_IS_EVENT_CONNECTED)
        if nEventConnected != None and nEventConnected == True:
            objRoot.newControlLoaded.disconnect(self.HandleNewControlLoadedEvent)
            objRoot.destroyControlLoaded.disconnect(self.HandleDestroyControlLoadedEvent)
            objRoot.setProperty(PROP_IS_EVENT_CONNECTED, False)

        self.SetControlEventDisconnection(objRoot, True)

    def SetControlEventDisconnection(self, obj, bRecursive):
        if not obj:
            getLogger().error('Object is null')
            return

        if bRecursive:
            self.SetControlEventDisconnectionRecursive(obj)
        else:
            self.disconnectObjectEvent(obj)

    def disconnectObjectEvent(self, obj):
        if not obj:
            getLogger().error('Object is null')
            return

        bConnectEvent = obj.property(PROP_B_CONNECT_EVENT)
        if not bConnectEvent:
            return

        nObjectType = obj.property('objectType')
        if not nObjectType:
            getLogger().error('Object type is null')
            return

        if nObjectType == ViewEnum.eObjectType.OBJECT_WINDOW:
            self.__hWindowEvent.disconnectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_VIEW:
            self.__hViewEvent.disconnectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_DIALOG:
            self.__hDlgEvent.disconnectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_BUTTON_CONTROL:
            self.__hBtnCtlEvent.disconnectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_COMBOBOX_CONTROL:
            self.__hCbxCtlEvent.disconnectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_TEXTEDIT_CONTROL:
            self.__hEdtCtlEvent.disconnectObjectEvent(obj)
        elif nObjectType == ViewEnum.eObjectType.OBJECT_LIST_VIEW:
            self.__hLstEvent.disconnectObjectEvent(obj)
        else: # nObjectType == ViewEnum::OBJECT_NONE or unknown
            getLogger().error('Unknown objectType: %d', nObjectType)

    def GetRootControlObject(self) -> QObject:
        objRoot = None

        if self.__engine == None:
            return objRoot

        lstRootObject = self.__engine.rootObjects()
        if len(lstRootObject) == 0:
            return objRoot

        for obj in lstRootObject:
            if obj != None:
                if obj.objectName() == ID_idRootWindow:
                    objRoot = obj
                    break

        return objRoot

    def FindChildObject(self, objectName:str, parent = None):
        if not objectName:
            return None

        if not parent:
            parent = self.GetRootControlObject()

        if not parent:
            return None

        if objectName == parent.objectName():
            return parent

        return parent.findChild(QtCore.QObject, objectName)

    def SetControlEventConnectionRecursive(self, obj):
        if not obj:
            getLogger().error('Object is null')
            return

        self.connectObjectEvent(obj)

        lstObjChild = obj.children()
        if len(lstObjChild) == 0:
            return

        for objChild in lstObjChild:
            if not objChild:
                continue

            self.SetControlEventConnectionRecursive(objChild)

    def SetControlEventDisconnectionRecursive(self, obj):
        if not obj:
            getLogger().error('Object is null')
            return

        self.disconnectObjectEvent(obj);

        lstObjChild = obj.children()
        if len(lstObjChild) == 0:
            return

        for objChild in lstObjChild:
            if not objChild:
                continue

            self.SetControlEventDisconnectionRecursive(objChild)

    @QtCore.pyqtSlot(str)
    def HandleNewControlLoadedEvent(self, objName:str):
        newObj = self.FindChildObject( objName )
        if not newObj:
            getLogger().error('Failed to get object: ', objName)
            return

        self.SetControlEventConnection(newObj, True)

    @QtCore.pyqtSlot(str)
    def HandleDestroyControlLoadedEvent(self, objName:str):
        newObj = self.FindChildObject( objName )
        if not newObj:
            getLogger().error('Failed to get object: ', objName)
            return

        self.SetControlEventDisconnection(newObj, True)

    @property
    def hWindowEvent(self):
        return self.__hWindowEvent

    @property
    def hViewEvent(self):
        return self.__hViewEvent

    @property
    def hLstEvent(self):
        return self.__hLstEvent

    @property
    def hBtnCtlEvent(self):
        return self.__hBtnCtlEvent

    @property
    def hCbxCtlEvent(self):
        return self.__hCbxCtlEvent

    @property
    def hDlgEvent(self):
        return self.__hDlgEvent

    @property
    def hEdtCtlEvent(self):
        return self.__hEdtCtlEvent

# See PyCharm help at https://www.jetbrains.com/help/pycharm/

File: QT/module/ViewControlModule.cpp
#include "ViewControlModule.h"

#include <QModuleManager.h>
#include <viewDefs.h>

#include <CLogger.h>

namespace nsViewControlModule{
    const char* TAG = "ViewControlModule";
}

using namespace nsViewControlModule;

ViewControlModule::ViewControlModule( QObject *parent )
    : QBaseModule(parent), m_pEngine(nullptr), m_pHWindowEvent(nullptr)
    , m_pHViewEvent(nullptr), m_pHLstEvent(nullptr), m_pHBtnCtlEvent(nullptr)
    , m_pHCbxCtlEvent(nullptr), m_pHDlgEvent(nullptr), m_pHEdtCtlEvent(nullptr)
{

}

ViewControlModule::~ViewControlModule()
{
    SetControlEventDisconnections();

    if( m_pEngine != nullptr )
    {
        m_pEngine->exit( 0 );
        m_pEngine->deleteLater();
        m_pEngine = nullptr;
    }
}

void ViewControlModule::SetControlEventConnections()
{
    QObject* pObjRoot = GetRootControlObject();
    assert( pObjRoot != nullptr );

    LOGD( TAG, "test" );

    QVariant vEventConnected = pObjRoot->property( PROP_IS_EVENT_CONNECTED );
    if( vEventConnected.isNull() || vEventConnected.isValid() == false
            || vEventConnected.toBool() == false )
    {
        connect( pObjRoot, SIGNAL( newControlLoaded(QString) ), this, SLOT( HandleNewControlLoadedEvent(QString) ), Qt::DirectConnection );
        connect( pObjRoot, SIGNAL( destroyControlLoaded(QString) ), this, SLOT( HandleDestroyControlLoadedEvent(QString) ), Qt::DirectConnection );
        pObjRoot->setProperty( PROP_IS_EVENT_CONNECTED, true );
        LOGD( TAG, "test" );
    }

    SetControlEventConnection( pObjRoot, true );
}

void ViewControlModule::SetControlEventConnection(QObject *pObj, bool bRecursive)
{
    if( pObj == nullptr )
    {
        LOGE( TAG, "Object is null" );
        return;
    }

    if( bRecursive )
        SetControlEventConnectionRecursive( pObj );
    else
        connectObjectEvent( pObj );
}

void ViewControlModule::connectObjectEvent(QObject *pObj)
{
    if( pObj == nullptr )
    {
        LOGE( TAG, "Object is null" );
        return;
    }

    QVariant vConnectEvent = pObj->property( PROP_B_CONNECT_EVENT );
    if( vConnectEvent.isNull() || vConnectEvent.isValid() == false
        || vConnectEvent.toBool() == false )
        return;

    QVariant vObjectType = pObj->property( "objectType" );
    if( vObjectType.isNull() || vObjectType.isValid() == false )
    {
        LOGE( TAG, "Object type is null" );
        return;
    }

    switch( vObjectType.toInt() )
    {
    case static_cast<int>( ViewEnum::OBJECT_WINDOW ):
        m_pHWindowEvent->connectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_VIEW ):
        m_pHViewEvent->connectObjectEvent( pObj );
    break;
    case static_cast<int>( ViewEnum::OBJECT_DIALOG ) :
        m_pHDlgEvent->connectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_BUTTON_CONTROL ):
        m_pHBtnCtlEvent->connectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_COMBOBOX_CONTROL ):
        m_pHCbxCtlEvent->connectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_TEXTEDIT_CONTROL ):
        m_pHEdtCtlEvent->connectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_LIST_VIEW ):
        m_pHLstEvent->connectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_NONE ) :
    default:
        LOGE( TAG, "Unknown objectType: %d", vObjectType.toInt() );
    break;
    }
}

void ViewControlModule::SetControlEventDisconnections()
{
    QObject* pObjRoot = GetRootControlObject();
    if( pObjRoot != nullptr )
    {
        QVariant vEventConnected = pObjRoot->property( PROP_IS_EVENT_CONNECTED );
        if( vEventConnected.isNull() == false && vEventConnected.isValid()
                || vEventConnected.toBool() )
        {
            disconnect( pObjRoot, SIGNAL( newControlLoaded(QString) ), this, SLOT( HandleNewControlLoadedEvent(QString) ) );
            disconnect( pObjRoot, SIGNAL( destroyControlLoaded(QString) ), this, SLOT( HandleDestroyControlLoadedEvent(QString) ) );
            pObjRoot->setProperty( PROP_IS_EVENT_CONNECTED, false );
        }

        SetControlEventDisconnection( pObjRoot, true );
    }
}

void ViewControlModule::SetControlEventDisconnection(QObject *pObj, bool bRecursive)
{
    if( pObj == nullptr )
    {
        LOGE( TAG, "Object is null" );
        return;
    }

    if( bRecursive )
        SetControlEventDisconnectionRecursive( pObj );
    else
        disconnectObjectEvent( pObj );
}

void ViewControlModule::disconnectObjectEvent(QObject *pObj)
{
    if( pObj == nullptr )
    {
        LOGE( TAG, "Object is null" );
        return;
    }

    QVariant vConnectEvent = pObj->property( PROP_B_CONNECT_EVENT );
    if( vConnectEvent.isNull() || vConnectEvent.isValid() == false
        || vConnectEvent.toBool() == false )
        return;

    QVariant vObjectType = pObj->property( "objectType" );
    if( vObjectType.isNull() || vObjectType.isValid() == false )
    {
        LOGE( TAG, "Object type is null" );
        return;
    }

    switch( vObjectType.toInt() )
    {
    case static_cast<int>( ViewEnum::OBJECT_WINDOW ):
        m_pHWindowEvent->disconnectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_VIEW ):
        m_pHViewEvent->disconnectObjectEvent( pObj );
    break;
    case static_cast<int>( ViewEnum::OBJECT_DIALOG ) :
        m_pHDlgEvent->disconnectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_BUTTON_CONTROL ):
        m_pHBtnCtlEvent->disconnectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_COMBOBOX_CONTROL ):
        m_pHCbxCtlEvent->disconnectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_TEXTEDIT_CONTROL ):
        m_pHEdtCtlEvent->disconnectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_LIST_VIEW ):
        m_pHLstEvent->disconnectObjectEvent( pObj );
        break;
    case static_cast<int>( ViewEnum::OBJECT_NONE ) :
    default:
        LOGE( TAG, "Unknown objectType: %d", vObjectType.toInt() );
    break;
    }
}

QObject* ViewControlModule::GetRootControlObject()
{
    QObject* pObjRoot = nullptr;

    if( m_pEngine == nullptr || m_pEngine->rootObjects().isEmpty() )
        return nullptr;

    QList<QObject*> lstRootObject = m_pEngine->rootObjects();

    for( QObject* pObj : lstRootObject )
    {
        if( pObj != nullptr )
        {
            if( pObj->objectName().compare( _idRootWindow ) == 0 )
            {
                pObjRoot = pObj;
                break;
            }
        }
    }

    return pObjRoot;
}

QObject* ViewControlModule::FindChildObject(const QString& objectName, QObject* pParent)
{
    if( objectName.isEmpty() )
        return nullptr;

    if( pParent == nullptr )
        pParent = GetRootControlObject();

    if( pParent == nullptr )
        return nullptr;

    if( objectName.compare( pParent->objectName() ) == 0 )
        return pParent;

    return pParent->findChild<QObject*>( objectName );
}

WindowEventHandler* ViewControlModule::GetWindowEventHandler()
{
    return m_pHWindowEvent;
}

ViewEventHandler* ViewControlModule::GetViewEventHandler()
{
    return m_pHViewEvent;
}

ListEventHandler* ViewControlModule::GetListEventHandler()
{
    return m_pHLstEvent;
}

ButtonEventHandler* ViewControlModule::GetButtonEventHandler()
{
    return m_pHBtnCtlEvent;
}

ComboBoxEventHandler* ViewControlModule::GetComboBoxEventHandler()
{
    return m_pHCbxCtlEvent;
}

DialogEventHandler* ViewControlModule::GetDialogEventHandler()
{
    return m_pHDlgEvent;
}

TextEditEventHandler* ViewControlModule::GetTextEditEventHandler()
{
    return m_pHEdtCtlEvent;
}

void ViewControlModule::SetControlEventConnectionRecursive(QObject *pObj)
{
    if( pObj == nullptr )
    {
        LOGE( TAG, "Object is null" );
        return;
    }

    connectObjectEvent( pObj );

    QList<QObject *> lstObjChild = pObj->findChildren<QObject*>( QString(), Qt::FindDirectChildrenOnly );
    if( lstObjChild.isEmpty() )
        return;

    for( QObject* pObjChild : lstObjChild )
    {
        if( pObjChild == nullptr )
            continue;

        SetControlEventConnectionRecursive( pObjChild );
    }
}

void ViewControlModule::SetControlEventDisconnectionRecursive(QObject *pObj)
{
    if( pObj == nullptr )
    {
        LOGE( TAG, "Object is null" );
        return;
    }

    disconnectObjectEvent( pObj );

    QList<QObject *> lstObjChild = pObj->findChildren<QObject*>( QString(), Qt::FindDirectChildrenOnly );
    if( lstObjChild.isEmpty() )
        return;

    for( QObject* pObjChild : lstObjChild )
    {
        if( pObjChild == nullptr )
            continue;

        SetControlEventDisconnectionRecursive( pObjChild );
    }
}

void ViewControlModule::init()
{
    if( initQmlEngine() == false )
    {
        LOGE( TAG, "Failed to init QML Engine" );
        return;
    }

    m_pHWindowEvent = new WindowEventHandler();
    m_pHViewEvent = new ViewEventHandler;
    m_pHLstEvent = new ListEventHandler;
    m_pHBtnCtlEvent = new ButtonEventHandler;
    m_pHCbxCtlEvent = new ComboBoxEventHandler;
    m_pHDlgEvent = new DialogEventHandler;
    m_pHEdtCtlEvent = new TextEditEventHandler;

    SetControlEventConnections();

    connect( m_pHEdtCtlEvent, SIGNAL( SigExcelSongInfoFilePathChanged(int, QString) )
             , this, SLOT( OnExcelSongInfoFilePathChanged(int, QString) ), Qt::DirectConnection );
    connect( m_pHEdtCtlEvent, SIGNAL( SigSrcPathChanged(int, QString) )
             , this, SLOT( OnSrcPathChanged(int, QString) ), Qt::DirectConnection );
    connect( m_pHEdtCtlEvent, SIGNAL( SigDstPathChanged(int, QString) )
             , this, SLOT( OnDstPathChanged(int, QString) ), Qt::DirectConnection );

    connect( m_pHBtnCtlEvent, SIGNAL( SigConvert(int) ), this, SLOT( OnConvert(int) ), Qt::DirectConnection );
    connect( m_pHBtnCtlEvent, SIGNAL( SigReload(int) ), this, SLOT( OnReload(int) ), Qt::DirectConnection );

    QBaseModule::init();
}

bool ViewControlModule::initQmlEngine()
{
    m_pEngine = new QQmlApplicationEngine();

    if( m_pEngine->rootContext() != nullptr )
    {
        m_pEngine->addImportPath( QStringLiteral("qrc:/") );
        m_pEngine->addImportPath( QStringLiteral("qrc:///BaseComponents") );
        m_pEngine->addImportPath( QStringLiteral("qrc:///components") );
    }

    qmlRegisterType<ViewEnum>( "ViewEnum", 1, 0, "ViewEnum" );

    const QUrl url(QStringLiteral("qrc:/main.qml"));

    QObject::connect(m_pEngine, &QQmlApplicationEngine::objectCreated
                     , this, [url](QObject *obj, const QUrl &objUrl) {
                            if (!obj && url == objUrl)
                                QCoreApplication::exit(-1);
                        }, Qt::QueuedConnection);

    m_pEngine->load(url);
    if( m_pEngine->rootObjects().isEmpty() )
        return false;

    return true;
}

QString ViewControlModule::getModuleName()
{
    return MDL_NAME( ViewControlModule );
}

void ViewControlModule::doRun()
{

}

void ViewControlModule::stopModule()
{
    QBaseModule::stopModule();

    clear();
}

void ViewControlModule::HandleNewControlLoadedEvent(QString objName)
{
    LOGD( TAG, "test object: %s", objName.toStdString().c_str() );
    QObject* pNewObj = FindChildObject( objName );
    if( pNewObj == nullptr )
    {
        LOGE( TAG, "Failed to get object: %s", objName.toStdString().c_str() );
        return;
    }

    SetControlEventConnection( pNewObj, true );
}

void ViewControlModule::HandleDestroyControlLoadedEvent(QString objName)
{
    LOGD( TAG, "test object: %s", objName.toStdString().c_str() );
    QObject* pNewObj = FindChildObject( objName );
    if( pNewObj == nullptr )
    {
        LOGE( TAG, "Failed to get object: %s", objName.toStdString().c_str() );
        return;
    }

    SetControlEventDisconnection( pNewObj, true );
}

void ViewControlModule::clear()
{

}
